docs/source/learn/trace-vs-execution-time.md

```diff
@@ -0,0 +1,110 @@
+
+When working with PyTorch/XLA, it's essential to understand that operations on
+XLA tensors are not typically executed immediately in the way they are with
+standard PyTorch tensors on CPU or CUDA devices (which operate in "eager mode").
+PyTorch/XLA employs a "lazy execution" model. This means that when you write
+PyTorch code using XLA tensors, you are primarily defining or tracing a
+computation graph. The compilation of the currently traced graph and it's
+subsequent execution on the device are deferred until a specific trigger point.
+
+This leads to two distinct types of "time" to consider:
+
+1. Host-Side Time: The period during which your CPU (host) prepares the
+computation. This includes:
+
+   - **Tracing Time**: The period during which PyTorch/XLA records your
+   operations and builds the computation graph.
+
+   - **Compilation Time**: The time the host-side XLA compiler takes to
+   transform the traced graph into optimized device code. This is most
+   significant on the first execution of a new graph or if the graph changes.
+
+2. Device Time: This is primarily the **Execution Time**, which is the period
+   during which the XLA device (e.g., TPU) spends running the compiled code.
+
+## Illustrating a Common Pitfall: Measuring Only Tracing Time
+
+When you write PyTorch code using XLA tensors (e.g., tensors on a TPU),
+PyTorch/XLA doesn't execute each operation on the device right away. It traces
+these operations, adding them to an internal computation graph. **If you measure
+the duration of code that only performs XLA operations without an explicit
+instruction to wait for the device, you are primarily measuring this tracing
+time plus Python overhead.**
+
+Consider the following conceptual code:
+
+```Python
+# Assume 'a' and 'b' are XLA tensors
+start_time = time.perf_counter()
+
+# This operation is recorded in PyTorch/XLA's graph
+result = torch.matmul(a, b)
+
+# ❌❌❌ !!! INCORRECT PROFILING: compilation and execution are deferred !!! ❌❌❌
+end_time = time.perf_counter()
+elapsed_time = end_time - start_time
+```
+
+The `elapsed_time` here would predominantly reflect how long it took PyTorch/XLA
+to trace the matmul operation. The actual matrix multiplication on the XLA
+device, along with its compilation, is not started.
+
+## Measuring End-to-End Performance
+
+To correctly profile the performance of your code on the XLA device, you must
+ensure that your timing captures host-side compilation and devide execution.
+This involves:
+
+1. Ensure the traced computational graph is compiled, if it's the first time
+this graph is seen or if it is changed, and sent to the device for execution.
+
+1. Make sure the Python script waits until the XLA device has completed all its
+assigned computations before taking the final timestamp.
+
+This is exemplified, using `torch_xla.sync(wait=True)`, in the following
+conceptual code:
+
+```Python
+# Assume 'a' and 'b' are XLA tensors
+start_time = time.perf_counter()
```

Comment:
nit: `start_time` should be measured after the first `sync(wait=True)`, and right before the second `sync(wait=True)`. This way we only measure the execution time and not tracing time.

Reply:
Ah I missed that. fixed.